### ISSUE TITLE
Conditions in color encodings are not reflected in legend

### ISSUE THREAD

**Issue by rlm1023** [vega/vega-lite#9497]
## Bug Description
If you have a color encoding with a conditional re-coloring on it for particular values, those re-colorings are not reflected in the legend.

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "data": {"url": "data/cars.json"},
  "mark": "point",
  "encoding": {
    "x": {"field": "Horsepower", "type": "quantitative"},
    "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
    "color": {
      "field": "Origin",
      "type": "nominal",
      "condition": {"test": "datum['Origin'] === 'USA'", "value": "red"},
      "scale": {"scheme": "category10"}
    }
  }
}
```

Produces:
![Image](https://github.com/user-attachments/assets/27f1546e-45fb-4b74-a63c-22434609bc35)

[Repro](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEySYUqUAwBOgtCrVICUJNohSZ8gL5Lk2gNb6cbGrMyKQcWVDbLXZNFAAD0CQADMaOEFlfQAJNjM4ZwB3OG0PTABPHDh9AEcGJDc6NRpSEBsQTNCIqJj0AFkaQTgIAH0c7TaAcSRBQTkM7Nz0AqLMEonyyp8B9M1wyOj9AHltGlp5JSyc-Vk2BFc+jx9ZPwnBhcxW93RVTAYEAG0AcjWN1xeAXQACAF4AT8XgBVADKAEEXh5iH0GCMQNo4DFKtA+iNQFxEPCTNcyAlMgBGAAMFSsZKAA)

<details>
  <summary><h2>Checklist</h2></summary>

- [x] I checked for duplicate issues.
</details>

**Comment by hydrosquall** [comment]
I see how this seems like a bug at first, but it may be a feature request as this isn't a regression, it was just not built before.

[Conditions](https://vega.github.io/vega-lite/docs/condition.html) were added independently of legends, and can be expressed in many forms. Since this form is using an expression test rather than a field predicate , Vega-lite is not readily aware that "origin" is being used in the test and that one of dots can be recolored.

I would suggest to think of the current legend as the color to use as long as no conditions are met. We could implement legends for conditional rules as a future enhancement, but it's currently not doable. 

If you wish to change the color for a specific country in a graph and have that reflected in the legend, you can do so with a custom (scale.scheme.range) instead of using a condition. Would that work for your case?

**Comment by rlm1023** [comment]
Yeah, I figure it might not be exactly a bug.

I've gotten around it so far by fully specifying the scale's domain/range, rather than using a condition. But that requires me to consume the entire dataset to build out those arrays. If there's a better way to do that, I'd be very curious how (I'm not familiar with scale.scheme.range, so if you have an example, I'd be thankful!).

**Comment by rlm1023** [comment]
I should also add that I've used (I think) the field predicate version of the condition, and seen the same issue.

[Repro](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEySYUqUAwBOgtCrVICUJNohSZ8gL5Lk2gNb6cbGrMyKQcWVDbLXZNFAAD0CQADMaOEFlfQAJNjM4ZwB3OG0PTABPHDh9AEcGJDc6NRpSEBsQTNCIqJj0AFkaQTgIAH0c7TaAcSRBQTkM7Nz0AqLMEonyyp8B9M1wyOj9AHltGlp5JSyc-Vk2BFc+jx9ZPwnBhcxW9wXa5fQ1jdcPODG9dABVAGUAQQqlMQ+gwRiBtHAYpVoH0RqAuIhQSZrmQEpkAIwABgqVhxQA)

**Comment by hydrosquall** [comment]
It sounds like you might already have tried this if you've already set a custom range, but just in case here's a custom [range example](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEySYUqUAwBOgtCrVICUJNohSZ8gL5Lk2gNb6cbGrMyKQcWVDbLXZNFAAD0CQADMaOEFlfQAJNjM4ZwB3OG0PTABPHDh9AEcGJDc6NRpSEBsQTNCIqJj0AFkaQTgIAH0c7TaAcSRBQTkM7Nz0AqLMEonyyp8B9M1wyOj9AHltGlp5JSyc-Vk2BFc+j2g+kdBtIrIRgG0QBKvcpSZBBieQYTIsMkvqgF0rICrEA)

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "data": {"url": "data/cars.json"},
  "mark": "point",
  "encoding": {
    "x": {"field": "Horsepower", "type": "quantitative"},
    "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
    "color": {"field": "Origin", "type": "nominal", "scale": {
      "range": [
        "orange",
        "blue",
        "lightgray"
      ]
    }}
  }
}
```

RE: field predicate: we don't use the conditions in any form for the legends, but thanks for linking that too. I made the point about the condition form mainly to say that if we were to implement it, it may be easier to label with the predicate form than the "arbitrary expression" form.

**Comment by rlm1023** [comment]
Yeah in that case it doesn't pin the color to specific values right? We're essentially using conditions as "these values should map to these colors, but the rest can come from the scheme".

**Comment by hydrosquall** [comment]
I think domain and range are linked (first domain value, first color), the
order matters. As long as you know all the domain values up front, you can
pin color to data values.

**Comment by rlm1023** [comment]
Yeah, that's actually what I wound up doing to fix legends. So we now define explicit domains / ranges. The problem with that is having to iterate large data sets to generate those arrays, and vega-lite seems to struggle when they get in to the thousands / tens of thousands.

But I can work with that for now.

**Comment by hydrosquall** [comment]
Thanks for the feedback - being able to set explicit colors for categories
and have those reflected in the legend is a reasonable use case.

For these large datasets, how many entries do you end up displaying in the
legend?

**Comment by rlm1023** [comment]
All of them, so thousands. This is likely not a realistic use case, as coloring by even 1000 values would likely not be a very useful visualization, and at that point we'd take it back to design. So in general, I'd expect the number to be generally small.

**Comment by hydrosquall** [comment]
Ah I see. I’m not sure how much screen space your graphs get, but the
legend size will be limited by the size of the graph surface since
currently the legends are not expandable.
